reconcile/utils/ldap_client.py

```diff
@@ -22,16 +20,20 @@ def init_from_config():
 
     serverUrl = config['ldap']['server']
     _base_dn = config['ldap']['base_dn']
-
     return init(serverUrl)
 
 
 def user_exists(username):
-    global _client
     global _base_dn
 
-    init_from_config()
+    client = init_from_config()
+
+    if not client.bound:
+        client.bind()
 
     search_filter = "uid={},{}".format(username, _base_dn)
+    result, _, _, _ = client.search(search_filter, '(objectclass=person)')
+
+    client.unbind()
 
-    return _client.search(search_filter, '(objectclass=person)')
+    return result
```

Comment:
previously we returned the result of client.search() now it looks like we're unpacking what it returns and only returning one of it's elements (result). that makes me wonder if the consumers of this function will still function?

or is it that the search() method return type changed in the new version?

can we add a test?

Reply:
The change in the output is a result of using a different client_strategy. Searches run with a connection parameter `client_strategy=SYNC` will return a single boolean value - did the search find a result or not? The actual found data is then accessible as additional parameters bound to the client object.

Moving to `client_strategy=SAFE_SYNC`, a tuple is returned instead of a boolean. The values that were written to the client object are instead returned to the caller.

The usage in this integration does not care about any of the data associated with a particular search. The only thing the integration cares about is if a user is present in ldap or not. If the user is not present, then the integration triggers a gitlab remove user method.

Reply:
The current usage of the ldap client is not particularly sophisticated - the integration simply checks with the ldap server to determine if a user record exists, but does nothing with the returned data and only operates on the boolean status variable. The thread safe connection returns a tuple instead of assigning found data to its object properties. Therefore, I ignore the additional return values and keep the integration (ldap_users.py) working as written.

Reply:
Re: testing, the only test I can think of would be to mock an ldap server and verify that we get correct booleans out of a user search for present / not present. This would duplicate tests in ldap3 package, no? Is there a different angle you can think of?